### ISSUE TITLE
Out of Memory Exceptions thrown during compilation

### ISSUE THREAD

**Issue by Porl91** [bridgedotnet/Bridge#3259]
We have about 8 Bridge projects currently in our main solution and recently we've been getting some intermittent `OutOfMemoryException` exceptions thrown when rebuilding. 

I'm not really sure how best to profile this to understand where the memory's going, although the stack traces do seem to be coming from the BridgeCompilerTask.

Here are some of the stack traces we've received recently:

<pre>System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at Mono.Cecil.PE.ByteBuffer.Grow(Int32 desired)
   at Mono.Cecil.PE.ByteBuffer.WriteBytes(Byte[] bytes)
   at Mono.Cecil.MetadataBuilder.AddResources()
   at Mono.Cecil.MetadataBuilder.BuildModule()
   at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__1_0(MetadataBuilder builder, MetadataReader _)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
   at Mono.Cecil.ModuleWriter.WriteModuleTo(ModuleDefinition module, Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
   at Bridge.Translator.Translator.EmbeddResources(List`1 resourcesToEmbed)
   at Bridge.Translator.Translator.InjectResources(String outputPath, String projectPath)
   at Bridge.Translator.TranslatorProcessor.PostProcess()
   at Bridge.Build.BridgeCompilerTask.Execute()</pre>

<pre>System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at Mono.Cecil.PE.ImageReader.ReadSectionData(Section section)
   at Mono.Cecil.PE.ImageReader.ReadSections(UInt16 count)
   at Mono.Cecil.PE.ImageReader.ReadImage()
   at Mono.Cecil.PE.ImageReader.ReadImageFrom(Stream stream)
   at Mono.Cecil.ModuleDefinition.ReadModule(Stream stream, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at Mono.Cecil.BaseAssemblyResolver.SearchDirectory(AssemblyNameReference name, IEnumerable`1 directories, ReaderParameters parameters)
   at Mono.Cecil.BaseAssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Bridge.Translator.Translator.CecilAssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Cecil.DefaultAssemblyResolver.Resolve(AssemblyNameReference name)
   at Bridge.Translator.Translator.CecilAssemblyResolver.Resolve(AssemblyNameReference name)
   at Bridge.Translator.Translator.GetParentAssemblies(AssemblyDefinition asm, List`1 list)
   at Bridge.Translator.Translator.GetParentAssemblies(AssemblyDefinition asm, List`1 list)
   at Bridge.Translator.Translator.SortReferences()
   at Bridge.Translator.Translator.Translate()
   at Bridge.Build.BridgeCompilerTask.Execute()</pre>

<pre>System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at Mono.Cecil.PE.ByteBuffer.Grow(Int32 desired)
   at Mono.Cecil.PE.ByteBuffer.WriteBytes(Byte[] bytes)
   at Mono.Cecil.MetadataBuilder.AddResources()
   at Mono.Cecil.MetadataBuilder.BuildModule()
   at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__1_0(MetadataBuilder builder, MetadataReader _)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
   at Mono.Cecil.ModuleWriter.WriteModuleTo(ModuleDefinition module, Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
   at Bridge.Translator.Translator.EmbeddResources(List`1 resourcesToEmbed)
   at Bridge.Translator.Translator.InjectResources(String outputPath, String projectPath)
   at Bridge.Translator.TranslatorProcessor.PostProcess()
   at Bridge.Build.BridgeCompilerTask.Execute()</pre>

<pre>System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.Collections.Generic.HashSet`1.SetCapacity(Int32 newSize, Boolean forceNewHashCodes)
   at System.Collections.Generic.HashSet`1.IncreaseCapacity()
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
   at Microsoft.Ajax.Utilities.JSVariableField.AddReference(INameReference reference)
   at Microsoft.Ajax.Utilities.ResolutionVisitor.ResolveLookup(ActivationObject scope, Lookup lookup, CodeSettings settings)
   at Microsoft.Ajax.Utilities.ResolutionVisitor.ResolveLookups(ActivationObject scope, CodeSettings settings)
   at Microsoft.Ajax.Utilities.ResolutionVisitor.ResolveLookups(ActivationObject scope, CodeSettings settings)
   at Microsoft.Ajax.Utilities.ResolutionVisitor.ResolveLookups(ActivationObject scope, CodeSettings settings)
   at Microsoft.Ajax.Utilities.ResolutionVisitor.Apply(Block block, ActivationObject scope, JSParser parser)
   at Microsoft.Ajax.Utilities.JSParser.InternalParse()
   at Microsoft.Ajax.Utilities.JSParser.Parse(DocumentContext sourceContext)
   at Microsoft.Ajax.Utilities.Minifier.MinifyJavaScript(String source, CodeSettings codeSettings)
   at Bridge.Translator.Translator.Minify(Minifier minifier, String source, CodeSettings settings)
   at Bridge.Translator.Translator.Minify(TranslatorOutputItem output, CodeSettings minifierSettings)
   at Bridge.Translator.Translator.Minify(IEnumerable`1 outputs, Func`2 minifierSettingsResolver)
   at Bridge.Translator.Translator.Minify()
   at Bridge.Translator.TranslatorProcessor.PostProcess()
   at Bridge.Build.BridgeCompilerTask.Execute()</pre>

Any ideas on what might be causing these, and also how we can mitigate them?

**Comment by DanTup** [comment]
It's worth noting that restarting VS seems to fix this issue temporarily. It normally happens after several builds.

Our projects are not very big (probably less than 50 classes each).

**Comment by DanTup** [comment]
Got this one during a clean build from the command line. Not sure if it was reusing any msbuild nodes though.

```text
C:\Users\DTuppeny\.nuget\packages\bridge.min\16.4.1\build\Bridge.Min.targets(22,5): 2017-11-07T14:57:06:842 	Error		System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.Text.StringBuilder.ExpandByABlock(Int32 minBlockCharCount)
   at System.Text.StringBuilder.Append(Char value, Int32 repeatCount)
   at System.Text.StringBuilder.Append(Char value)
   at System.IO.StringWriter.Write(Char value)
   at Microsoft.Ajax.Utilities.OutputVisitor.WriteToStream(Char ch)
   at Microsoft.Ajax.Utilities.OutputVisitor.OutputPossibleLineBreak(Char ch)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(AstNodeList node)
   at Microsoft.Ajax.Utilities.AstNodeList.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(ObjectLiteral node)
   at Microsoft.Ajax.Utilities.ObjectLiteral.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.AcceptNodeWithParens(AstNode node, Boolean needsParens)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(ArrayLiteral node)
   at Microsoft.Ajax.Utilities.ArrayLiteral.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.AcceptNodeWithParens(AstNode node, Boolean needsParens)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(ObjectLiteralProperty node)
   at Microsoft.Ajax.Utilities.ObjectLiteralProperty.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(AstNodeList node)
   at Microsoft.Ajax.Utilities.AstNodeList.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(ObjectLiteral node)
   at Microsoft.Ajax.Utilities.ObjectLiteral.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(ReturnNode node)
   at Microsoft.Ajax.Utilities.ReturnNode.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(Block node)
   at Microsoft.Ajax.Utilities.Block.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.OutputFunctionArgsAndBody(FunctionObject node)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(FunctionObject node)
   at Microsoft.Ajax.Utilities.FunctionObject.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.AcceptNodeWithParens(AstNode node, Boolean needsParens)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(CallNode node)
   at Microsoft.Ajax.Utilities.CallNode.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(Block node)
   at Microsoft.Ajax.Utilities.Block.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.OutputFunctionArgsAndBody(FunctionObject node)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(FunctionObject node)
   at Microsoft.Ajax.Utilities.FunctionObject.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.AcceptNodeWithParens(AstNode node, Boolean needsParens)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(CallNode node)
   at Microsoft.Ajax.Utilities.CallNode.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Visit(Block node)
   at Microsoft.Ajax.Utilities.Block.Accept(IVisitor visitor)
   at Microsoft.Ajax.Utilities.OutputVisitor.Apply(TextWriter writer, AstNode node, CodeSettings settings)
   at Microsoft.Ajax.Utilities.Minifier.MinifyJavaScript(String source, CodeSettings codeSettings)
   at Bridge.Translator.Translator.Minify(Minifier minifier, String source, CodeSettings settings)
   at Bridge.Translator.Translator.Minify(TranslatorOutputItem output, CodeSettings minifierSettings)
   at Bridge.Translator.Translator.Minify(IEnumerable`1 outputs, Func`2 minifierSettingsResolver)
   at Bridge.Translator.Translator.Minify()
   at Bridge.Translator.TranslatorProcessor.PostProcess()
   at Bridge.Build.BridgeCompilerTask.Execute() [C:\Work\Source\DMS\DMS-3-17-11\CreateMailoutTask.Tests\CreateMailoutTask.Tests.csproj]
```

I ran it again with resource monitor open and the memory did creep up to several GB for the MSBuild processes (though didn't get as far as crashing).

![memory](https://user-images.githubusercontent.com/1078012/32500158-70ecab0a-c3cc-11e7-8342-7d47d781dc0f.jpg)

**Comment by vladsch** [comment]
I never run into such issue. Is it reproducible if disable minification in bridge.json?

**Comment by DanTup** [comment]
It just occurred to me that this process is likely being shared with non-bridge projects, so the size of our solution (around 80 projects) is likely a factor. I'll try and do some more testing when I get time to see what the memory usage pattern is like if we remove all of the Bridge projects to see whether Bridge is actually using a lot, or whether it's just the last straw!

**Comment by fabriciomurta** [comment]
While building full bridge with test projects and all I can notice a Bridge.exe process taking ~700MB and at another time another is spawned going up to ~1.2GB (rounding up). But when they quit, all becomes clean again. [Bridge tests](https://testing.bridge.net/) is quite expensive in regards to code, I believe, then that should have been a good example.

Watching the build in process explorer, I can see the base memory footprint (11.9GB) goes up to 13.6GB during the build, But again, I don't see this number viciously increasing over time. I've done several rebuilds without closing+reopening VS. Sometimes some processes gain 2MB in memory allocation, but then later they show up 2MB behind my reference value, so I'm not getting any leaks here.

Using VS2017 15.1 (26403.7) here.

**Comment by fabriciomurta** [comment]
https://stackoverflow.com/questions/6107322/memory-limitations-in-a-64-bit-net-application
-maybe- switching the project to .NET 4.5 helps with the memory limitation. You should be able to link your 4.5 project to bridge. It would help if the issue could be consistently reproduced, though.

**Comment by DanTup** [comment]
I'll try and do some more investigation when back in the office and see if I can figure a pattern to reliably repro, and also what reduction we see if we don't build the Bridge projects. We're already targeting .NET 4.5 for our projects, but presumably the comments about that relate to the running code (eg. MSBuild/Bridge compiler code)?

I'll post back with any more info I can gather soon.

**Comment by fabriciomurta** [comment]
If we can identify the problem is a scenario where bridge breaches the 2GB object limit then we will be able to envise whether it is enough the project to have it set up or specific Bridge assemblies -- and how to bundle that appconfig setting to lift the limit (see the stackoverflow question's answers).

**Comment by andersnm** [comment]
This commit in Mono.Cecil addresses a leak, and was released in Mono.Cecil 0.10.0-beta7, alas wasn't straight-forward to swap out in Bridge since there are other breakages:

https://github.com/jbevain/cecil/commit/1b5ce064bb1661312bf8ade3c2d736a9fe452db5

... haven't been able to test if this changes anything yet

**Comment by DanTup** [comment]
We're seeing this *really* frequently now (every other build on our buildserver fails). @fabriciomurta is it likely Cecil can be upgraded to see whether the fix linked by @andersnm solves the problem for us?

**Comment by andersnm** [comment]
Had another go at this, and have a version of Bridge linked against the latest Cecil 0.10.0-beta7 and a custom NRefactory. 

But after the upgrade, am consistently getting the following error when building Bridge during the "Bridge.PostBuild" step:
```
5>  2017-11-29T13:21:51:579 	Error		Bridge.NET Compiler error: System.IO.IOException: The process cannot access the file 'C:\Users\anders\Code\Bridger\Bridge\bin\Release\Bridge.dll' because it is being used by another process.
5>     at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
5>     at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
5>     at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
5>     at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
5>     at Bridge.Translator.Translator.EmbeddResources(List`1 resourcesToEmbed) in C:\Users\anders\Code\Bridger\Compiler\Translator\Translator\Translator.Resources.cs:line 478
5>     at Bridge.Translator.Translator.InjectResources(String outputPath, String projectPath) in C:\Users\anders\Code\Bridger\Compiler\Translator\Translator\Translator.Resources.cs:line 63
5>     at Bridge.Translator.TranslatorProcessor.PostProcess() in C:\Users\anders\Code\Bridger\Compiler\Translator\Translator\TranslatorProcessor.cs:line 101
5>     at Bridge.Builder.Program.Main(String[] args) in C:\Users\anders\Code\Bridger\Compiler\Builder\Program.cs:line 46
```

However, the Bridge solution does successfully produce a Bridge.Min nupkg, and when hotswapping the Bridge-binaries used to build our solution, and building, our Bridge projects produce the same exception.

Looks like there were additional changes in Cecil introducing further leaks in Bridge. And Bridge does not appear to dispose any disposable objects. A quick attempt at implementing+invoking Translator.Close() did not yield positive results.

**Comment by DanTup** [comment]
I've been trying to run the build through ANTS memory profiler, but every time the memory rockets (which it does very steeply at certain parts of the build) and I click "Take Snapshot" (which forces a full GC then takes a snapshot) the memory immediately drops. This means the snapshot then doesn't show much memory being used at all.

I'm not sure how to interpret this - does this suggest there are a lot of objects that have been around long enough to be promoted, but actually are no longer required and therefore are available for collection?

Maybe there are appropriate places to force a full GC during the build (eg. if there's some early stage that allocates a ton of objects which are then no longer needed; maybe they've ended up in higher generations)?

I'll try and figure out if there are ways to get get ANTS to take a snapshot without forcing a full collection.

**Comment by DanTup** [comment]
I found I can add counters for the size of each generations heap...  However none of them make any second to me (none of them show any increase in any heap sizes for any processes that match the increase in the Private Bytes counter).

MSBuild's Private Bytes climbs up to around 1GB and VBCSCompiler's climbs up to around 2GB. The only heap size counter that shows anything significant is "Gen 0 heap size" for VBCSCompiler, but it's high from the start and barely changes throughout the whole compilation.

In the VBCSCompiler process there is a huge amount of "unused memory allocated to .NET" (862MB) but since that's right after a full collection, I don't know if that's because a lot has just been freed (the generation heap sizes don't seem to suggest so, but I must admit I don't know what I'm doing 😆)

**Comment by DanTup** [comment]
This is what I see in ANTS for the MSBuild process:

![MSBuild](https://user-images.githubusercontent.com/1078012/33381576-53d596d4-d516-11e7-8a7b-85bd0d7260f2.jpg)

The thick green line is "Private Bytes". The vertical lines are snapshots - the drop in memory slightly before them is the GC caused by taking a snapshot. All other counters (top levft) like Gen 0 heap size, Promoted Bytes, Allocated Bytes are just flat lines across the bottom (so according to these, no heaps are using the memory).

At the time of the snapshots (after the GC) the memory usage is fairly low, and therefore the bottom half of the screen is somewhat useless.

Here's the same for VBCSCompiler. This one does have some big spikes in allocations (think blue line) but most memory is used by Gen 0 heap (orange line) until the 4th snapshot when that starts to go down. This process only hits about half the usage of the MSBuild one (I don't know which one is running the code in the stack trace for the OOMs we're seeing).

![VBCSCompiler](https://user-images.githubusercontent.com/1078012/33381780-de025982-d516-11e7-8852-6fcda345251c.jpg)

**Comment by DanTup** [comment]
This is a bit old, but more discussion about Cecil memory usage:

**Comment by andersnm** [comment]
Have reproduced the OutOfMemoryException consistently in a solution on every rebuild: 

Basically, I copied all our Bridge projects into a new solution, and duplicated one of them a couple times, so there are totally 7 Bridge projects. There is ca 1mb of C# source code across the projects, dependencies not included. 

Because most of these projects have no interdependencies, they build in parallell! And it turns out the exception disappears after disabling parallel builds in VS.

Workaround:

Tools -> Options -> Projects and solutions -> Build and run -> Maximum number of parallel project builds -> Set to 1

**Comment by fabriciomurta** [comment]
Well, this result suggests not a memory leak in cecil, nor bridge, maybe. But just stacked calls to the Bridge builder DLL when using the targets' build task.

So, you probably could mitigate this issue, and still build in parallel if you switched to bridge executable call based build. Historically, the msbuild task has been source of visual studio-constrained issues with Bridge, like @andersnm shown [here](https://github.com/bridgedotnet/Bridge/issues/3259#issuecomment-347847295), by visual studio locking up the bridge DLL file after the first build.

If you manipulate your [Bridge.Min.targets](https://github.com/bridgedotnet/Bridge/blob/a9ac129ff46ae6d555ae008d2df6d96f62fb7a68/.build/files/Bridge.Min.targets) to use *BridgeCompilerUtil* instead of *BridgeCompilerTask* (the file is located within the Bridge.Min NuGet package, under the *build/* directory), then I believe you'll be lifting the memory limit by calling individual Bridge processes in parallel instead of stacking up in the DLL task call.

**Comment by andersnm** [comment]
Made myself a version of Bridge which prints GC.GetTotalMemory() before and after the build; i.e how much the GC believes has been allocated before and after compilation. 

Our Bridge projects use the following amounts of RAM (in bytes) while compiling:
```
Project "D":
Before:  7 695 464
After: 426 066 600

Project "ML":
Before:   15 934 396
After: 1 120 627 056

Project "TE":
Before:    7 599 552
After: 1 180 793 912

Project "CM":
Before:   12 165 032
After: 1 635 174 736

Project "EP":
Before:   15 901 140
After: 1 200 764 620

Project "P":
Before: 10 298 952
After: 873 593 792
```

F.ex project "ML" has ~150k of source files, and requires a gig of RAM to compile. If I strip down the "ML" project so there are zero files, Bridge reports consuming 262megabytes of RAM at the end. As I'm adding files back to see what actually consumes memory, it does not follow by the amount of code, but seems to make big jumps whenever referencing new types from externally referenced assemblies. 

F.ex adding an empty class to the otherwise empty project, and memory use stays pretty much the same. Then add a member to it using a type from an external assembly: `public TextBoxModel Text { get; set; }`, and the memory usage jumps from 262meg to 400meg. Am getting similar jumps when adding references to new types from the same assembly. Having multiple properties of type TextBoxModel does not seem to affect the memory usage.

The TextBoxModel type is relatively simple, so I can copy it into the same project and use it directly. Memory usage is then 300mb. 

But the TextBoxModel has a member of type `ProductiveRage.Immutable.Optional<ProductiveRage.Immutable.NonBlankTrimmedString>` - also in an external dependency, so I pulled in all the relevant source code from ProductiveRage.Immutable into my project. Memory usage is then 268meg! 

So it appears the same code in different assemblies = uses a lot of memory, and same code in single assembly = uses less memory.

**Comment by DanTup** [comment]
@andersnm Could the usage be relative to the size of the output JS? Many of our stack traces seem to be in minification, reading/writing embedded resources etc.. Pulling in external libs presumably causes all of their embedded JS files to be read and written back into the referencing library?

**Comment by andersnm** [comment]
Cannot easily test this anymore (reverted some stuff), but the tests did log GC.GetTotalMemory() before and after embedding/minification in the "PostProcess()" method. However the values were relatively close, so I didn't investigate that angle much. Also GC.GetTotalMemory collects, so any memory usage from embedding/minification comes in addition to the above numbers - probably what breaks the camel's back.

**Comment by DanTup** [comment]
@fabriciomurta is there enough info here for you to investigate whether the memory usage can be reduced? Changing the targets file as you suggest is non-trivial (since it's in package caches across 10 dev machines and build server; we'd end up having to maintain our own packages), but I also worry it's just putting things off a little longer (since it seems like our big projects may hit limits on their own in time).

**Comment by geoffreymcgill** [comment]
@andersnm & @DanTup Thanks for additional details. We will investigate and let you know if we have any questions.

**Comment by andersnm** [comment]
Most of our projects reference around 50+ Bridge DLLs directly and indirectly - 40 of those are custom UI components each in their own DLL. These files occupy ca 15mb in the bin/Debug folder after a build.

Have created a solution from scratch with projects "Project1" through "Project6", each containing a single class "Type1" through "Type6" with no members. 
"Project1" has project references to "Project2" through "Project6". 
"Project5" and "Project6" have a package reference to ProductiveRage.Immutable - "Type5" and "Type6" implement IAmImmutable with no members.

Then, using the modified Bridge compiler, rebuild "Project1" iteratively while adding back members with types from the other projects. These are the numbers I saw:

```c#
public class Type1
{
	// public Type2 Value2 { get; set; }
	// public Type2b Value2b { get; set; }
	// public Type3 Value3 { get; set; }
	// public Type4 Value4 { get; set; }
	// public Type5 Value5 { get; set; }
	// public Type6 Value6 { get; set; }

	// Uncomment one property at a time, rebuild, and note the memory usage logged by the Bridge compiler for Project1:
	// 104390112 (initial, all members in Type1 commented out)
	// 113740148 (Type2 uncommented)
	// 113752496 (Type2b uncommented, another empty class in Project2)
	// 165017048 (Type3 uncommented, embeds 2x10mb data)
	// 174363700 (Type4 uncommented)
	// 205644772 (Type5 uncommented, implements IAmImmutable)
	// 225180372 (Type6 uncommented, implements IAmImmutable)
}
```

Compared to an empty program, the memory usage increases by 120mb during compilation after adding 6 members referencing 6 empty classes in 5 assemblies, one assembly having 2x10mb embedded data.

The first reference to a new assembly allocates at minimum ca 10mb memory. (Type2, Type4)
Adding second reference to same assembly allocates no additional memory. (Type2b)
A reference with 20mb embedded binaries allocates 50mb (10+2x20mb) memory (Type3)
A reference implementing IAmImmutable allocates 30mb (Type5) or 20mb (Type6)

Following this pattern, a project involving 50 assemblies should probably expect pushing the 2gig limit soon, even with little code. 

---

Modified Bridge compiler:

```c#
        // Compiler\Translator\Translator\TranslatorProcessor.cs

        public string PostProcess()
        {
			...
            logger.Info("PostProcess end GC.GetTotalMemory " + GC.GetTotalMemory(true));

            return outputPath;
        }

```

Zipped VS2017 solution with "Project1".."Project6": [BigBridge.zip](https://github.com/bridgedotnet/Bridge/files/1523751/BigBridge.zip)

Download "Small File (10MB)" from https://www.thinkbroadband.com/download and save under Project3\10mb.zip in the unpacked directory

**Comment by DanTup** [comment]
Both teams here are now seeing this issue on their (distinct) codebases. While my teams project is very large (80 projects - most of which are not Bridge), the other teams solution is less than 10 projects (all new work, probably about half Bridge).

**Comment by andersnm** [comment]
In the method `Translator.GetParentAssemblies()`, called from `Translator.SortReferences()`:

https://github.com/bridgedotnet/Bridge/blob/a9ac129ff46ae6d555ae008d2df6d96f62fb7a68/Compiler/Translator/Translator/Translator.cs#L224

GetParentAssemblies() loops recursively through all assemblies in the project, starting from the main assembly like so:

```c#
	foreach (var assemblyReferenceName in asm.MainModule.AssemblyReferences)
	{
		...
		var assemblyReference = asm.MainModule.AssemblyResolver.Resolve(assemblyReferenceName);
		...
		GetParentAssemblies(assemblyReference, list);
	}

```

A "leak" occurs in the line with `AssemblyResolver.Resolve()`: Each AssemblyResolver instance maintains its own cache of assemblies (implemented in Cecil/[DefaultAssemblyResolver](https://github.com/jbevain/cecil/blob/0.9.6/Mono.Cecil/DefaultAssemblyResolver.cs)), and does not reuse/share previously loaded AssemblyDefinition instances from other resolvers. E.g when two assemblies reference the same dll, that dll is read into memory twice.

Rewritten as follows works better:

```c#

	foreach (var assemblyReferenceName in asm.MainModule.AssemblyReferences)
	{
		...
		var assemblyReference = References.Where(r => r.FullName == assemblyReferenceName.FullName).FirstOrDefault();
		if (assemblyReference == null)
			assemblyReference = asm.MainModule.AssemblyResolver.Resolve(assemblyReferenceName);
		...
		GetParentAssemblies(assemblyReference, list);
	}
```

After this change there is no longer an increase of memory usage during `SortReferences()` in my test. The "BigBridge" solution mentioned earlier went from 225mb to 160mb in memory usage during compilation.

**Comment by andersnm** [comment]
Have tested to build our main solution with the suggested fix, and it definitely helps. F.ex the project "ML" now uses minimum 600mb ram while compiling, down from 1100mb. Project "CM" uses 900mb down from 1600mb.

There are likely more "leaks" like this, in particular when indirectly referencing multiple/different versions of an assembly:

F.ex project "ML" uses Bridge 16.4.1 and references UI component DLLs built against Bridge 16.3.0. So SortReferences() still loads an additional copy of Bridge.dll from 16.3.0 into memory for every UI component. That's a minimum of 10mb waste per reference. With 40 references, that's 400mb. If addressed, the expected memory usage to build "ML" should be around ~200mb (instead of 1100mb currently)

Note that if the initially suggested fix is applied to Bridge, we can normalize the Bridge versions across our projects to address that last issue ourselves.

**Comment by fabriciomurta** [comment]
Just created the pull request above in branch *Issue3259* based (well, heavily) on your suggestion. I don't have myself a test scenario where I see difference in memory allocation by then but you'll probably see this in your project if the implementation here is right. But definitely your change suggestion does not break anything -- that is, if I didn't miss the point in the way I merged the change in the affected method.

**Comment by andersnm** [comment]
To repro, here is a zip file with two solutions for VS2017 throwing OutOfMemoryException during build: [Scenarios.zip](https://github.com/bridgedotnet/Bridge/files/1549011/Scenarios.zip)

- `Scenario1\Scenario.sln` - builds only after the fix in #3305 is applied to %USERPROFILE%\\.nuget\packages\bridge.min\16.4.1\tools\
- `Scenario2\Scenario.sln`  - throws OutOfMemoryException still

Both solutions contain 250 empty classes in 250 separate projects and one "Master" project which references them all. Thus the expected memory usage is at least 251*10mb (2.5gig!). 

The only difference between them is Scenario1 uses Bridge 16.4.1 for all projects, whereas Scenario2 uses Bridge 16.5.0 only for the "Master" project and Bridge 16.4.1 for the rest.

Scenario2 demonstrates how version mismatches in indirect assembly references cause similar leaks still.

And here is the project used to generate the above solutions: [BridgeBang.zip](https://github.com/bridgedotnet/Bridge/files/1549022/BridgeBang.zip)

**Comment by fabriciomurta** [comment]
Hey, thanks for sharing the bridge bang theory! I could run it and it crashes here with the out of memory exception just as it is supposed to! I will try to reduce the problem furthermore and see what else can be improved besides the suggestion you shared.

So you applied the fix for both 16.4.1 and 16.5.0, or you just replaced the DLLs within (thus making the 16.4.1\tools contain 16.5.0 (or 16.5.1) files?

**Comment by andersnm** [comment]
Yes, I built the master branch from Github w/the fix, and unpacked the \tools folder from the built Bridge.min.nupkg manually into the 16.4.1\tools directory to test it.

Imho a proper fix involves overhauling how Bridge deals with references: SortReferences() shouldn't need to resolve anything, afaict all references should've been resolved already during InspectReferences().

**Comment by andersnm** [comment]
The fix should probably look up previously resolved references by `Name` rather than `FullName`, and throw otherwise. Execution should never reach `SortReferences()` unless all references were resolved already, and shouldn't care if the resolved version matches exactly.

 Fixed fix looks like: (no overhaul needed!)

```c#
	foreach (var assemblyReferenceName in asm.MainModule.AssemblyReferences)
	{
		...
                // Changed r.FullName to .Name.Name and assemblyReferenceName.FullName to .Name:
		var assemblyReference = References.Where(r => r.Name.Name == assemblyReferenceName.Name).FirstOrDefault();
		if (assemblyReference == null)
			throw new InvalidOperationException(assemblyReferenceName.Name + " wasn't resolved");
		...
		GetParentAssemblies(assemblyReference, list);
	}
```

Pretty sure this is it. Scenario2 builds with this fix too.

**Comment by fabriciomurta** [comment]
Then it makes sense your scenario2 didn't fix. Your 16.4.1 is equivalent to current 16.5.1 (latest/development, with the fix). And your 16.5.0 would be "stock" (NuGet) version, without the fix.

You built the individual projects using the patched/fixed version whereas you tried to build the whole thing (main project) with an unpatched bridge version. Effectively the fixed translator run where it wouldn't demand the fix and the non-fixed translator ran where the issue would trigger.

I have just tested with #3305 increasing the size to 500 referenced projects, it does not break out of memory. Seems the number of projects no longer poses an issue as it just don't keep re-resolving the references. But even with the fix I noticed bridge gets as high as 800-1Gb RAM usage while building bridge tests (https://testing.bridge.net/) but that's quite a big project per se, not a big referencing network.

**Comment by andersnm** [comment]
I did update my 16.5.0 binaries with the fix in #3305 to verify Scenario2 still throws OutOfMemoryException. Then I applied the updated fix with .Name and verified Scenario2 no longer throws OutOfMemoryException. This is because the version using .Name reuses the main Bridge.dll, instead of loading the old version into memory 250 times.

**Comment by fabriciomurta** [comment]
> The fix should probably look up previously resolved references by Name rather than FullName,

I first thought I should reduce the test from FullName to something smaller or faster, but this part does not run a lot of times during build and also there is a chance a referenced assembly references an older version of an already resolved assembly, in which case it should try to resolve it and break if that version is not available, unless I'm missing the point here.

I'm still not sure your second version of the fix is safe, I don't see how the non-fixed "main" project should work just by having the sub-projects pointing to the fixed revision. Probably @vladsch can confirm your second solution works best.

**Comment by fabriciomurta** [comment]
> and throw otherwise

That "thow" in line 7 would totally trigger for the tests I ran while implementing the fix currently in #3305. Well, it's worth double checking, but I'm positive it needed to resolve parent assemblies in some situations I caught during my tests here.

**Comment by andersnm** [comment]
As expected, with the updated fix project "ML" uses minimum 200mb ram to build, down from 1100mb, and project "CM" uses 265mb, down from 1600mb.

Imo the `throw` is fine when comparing by the version-independent `.Name`. The assembly being looked up is guaranteed to have been resolved earlier, but not necessarily the exactly referenced version. Check out the implementation of `Translator.LoadAssembly()` where the initial resolving happens:

https://github.com/bridgedotnet/Bridge/blob/a9ac129ff46ae6d555ae008d2df6d96f62fb7a68/Compiler/Translator/Translator/Translator.InspectAssembly.cs#L87

**Comment by andersnm** [comment]
Unless I'm not missing something very subtly obvious, I believe it's possible to simplify `SortReferences()` as follows: https://github.com/andersnm/Bridge/commit/112a06842e99c1f10d1be63d5f73d7d381f47a8f

This builds a simpler dependency graph (without leaks) and afaict still produces a valid sort order.

It is a remix of the current sorting code which was recently implemented: https://github.com/bridgedotnet/Bridge/pull/2718

**Comment by fabriciomurta** [comment]
The simplification you proposed looks like a neat alternative for the fix (as it eliminates the trouble method completely) and we'd love to test it if you submitted us a pull request with it!

**Comment by fabriciomurta** [comment]
The simplification of SortReferences() was merged upstream and will be included in Bridge 16.6.0 (or whatever milestone this issue is bound to). This fixes at least the scenarios provided by @andersnm, and hopefully will address the scenarios that led to out of memory errors to @DanTup and @Porl91 as well.

Thanks for the awesome fix, @andersnm!

**Comment by DanTup** [comment]
The projects @andersnm was testing with are the same that Paul and I were working on, so this fix covers them 🙂

**Comment by DanTup** [comment]
@fabriciomurta @geoffreymcgill Is there an ETA for when this version will go out? If it's a way off, we might consider building a custom version for our local NuGet feed, as it's really frustrating some devs :(

(alternatively, any possibility of a beta release with it? We don't need to be able to ship for a while, just want to fix dev issues).

**Comment by fabriciomurta** [comment]
We want to release it this week, if all goes well.

**Comment by DanTup** [comment]
Sounds good; thanks!

**Comment by DanTup** [comment]
This morning I mentioned in our standup that we hadn't seen this since upgrading... However, I spoke too soon :(

```text
C:\Users\CCNet\.nuget\packages\bridge.min\16.6.0\build\Bridge.Min.targets (23,5):
error: 2018-01-04T10:07:07:439 	Error		System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at Mono.Cecil.PE.ByteBuffer.Grow(Int32 desired)
```
